### PR TITLE
Add Equal Loudness Filter

### PR DESCRIPTION
**Pull request by Martmists-GH**
### Describe your change:

Add an Equal Loudness Filter which aims to compensate for the human ear's non-linear response to sound. Based on the matlab implementation from [hydrogenaudio (archive.org link)](     https://web.archive.org/web/20130119143147/http://replaygain.hydrogenaudio.org/proposal/equal_loudness.html)

Unfortunately the precommit hook really wants to format the frequencies and gains list in a very specific way, making it look awful.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: audio_filters/equal_loudness_filter.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+from json import load
+from os.path import abspath, dirname
```
Comment (cclauss) [inline audio_filters/equal_loudness_filter.py:2]: Let's modernize...
```suggestion
from pathlib import Path
```

File: audio_filters/equal_loudness_filter.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+import numpy as np
+from yulewalker import yulewalk
+
+from audio_filters.butterworth_filter import make_highpass
+from audio_filters.iir_filter import IIRFilter
+
+with open(abspath(dirname(__file__)) + "/loudness_curve.json") as fp:
+    data = load(fp)
```
Comment (cclauss) [inline audio_filters/equal_loudness_filter.py:11]:
```suggestion
data = load(Path("loudness_curve.json").read_text())
```

File: audio_filters/loudness_curve.json [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    79,
+    84,
+    86,
+    86,
+    85,
+    95,
+    110,
+    125
+  ]
+}
```
Comment (cclauss) [inline audio_filters/loudness_curve.json:76]:
```suggestion
}

```

File: audio_filters/equal_loudness_filter.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        self.yulewalk_filter.set_coefficients(ya, yb)
+
+    def process(self, sample: float) -> float:
+        """
+        Process a single sample through both filters
+
+        >>> filt = EqualLoudnessFilter()
+        >>> filt.process(0.0)
+        0.0
+        """
```
Comment (cclauss) [inline audio_filters/equal_loudness_filter.py:60]: Could we please have some more difficult tests?  `.process(6)`, `.process(-6)`, `.process(-12345.6789)`

File: requirements.txt [hunk 5]
```diff
@@ -17,3 +17,4 @@ tensorflow
 texttable
 tweepy
 xgboost
+yulewalker
```
Comment (cclauss) [inline requirements.txt:20]: Normally I am reluctant to add dependencies but let's go for it.

**Review by cclauss — CHANGES_REQUESTED** [review]

File: audio_filters/equal_loudness_filter.py (hunk 4 above)
Comment (Martmists-GH) [reply audio_filters/equal_loudness_filter.py:60]: I simply copied the behavior from iir_filter.py, and the best way to test filters like these is to visually look at `show_frequency_response(my_filter)` but as it's matplotlib-based it's hard to check against in doctests.

File: requirements.txt (hunk 5 above)
Comment (Martmists-GH) [reply requirements.txt:20]: I'd love to use scipy's function, but at the moment it's [just a stub](https://github.com/scipy/scipy/blob/ee9985aafef8d3d99e19b2594767cf5fb8bfc3e6/scipy/signal/_filter_design.py#L3621)

File: audio_filters/equal_loudness_filter.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        #  fit to the specified curve.
+        ya, yb = yulewalk(10, freqs_normalized, gains_normalized)
+        self.yulewalk_filter.set_coefficients(ya, yb)
+
+    def process(self, sample: float) -> float:
+        """
+        Process a single sample through both filters
+
+        >>> filt = EqualLoudnessFilter()
+        >>> filt.process(0.0)
```
[the remark this answers is not in the document]
Comment (cclauss) [reply audio_filters/equal_loudness_filter.py:58]: No.  Stay focused on this file.

**Comment by Martmists-GH** [comment]
@cclauss I applied your suggestions, would you happen to know why it's not working?

File: audio_filters/equal_loudness_filter.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
+from json import load
+from pathlib import Path
+
+import numpy as np
+from yulewalker import yulewalk
+
+from audio_filters.butterworth_filter import make_highpass
+from audio_filters.iir_filter import IIRFilter
+
+data = load(Path("loudness_curve.json").read_text())
```
Comment (cclauss) [inline audio_filters/equal_loudness_filter.py:10]:
```suggestion
data = load(Path(__file__).resolve().parent / "loudness_curve.json").read_text())
```

**Review by cclauss — APPROVED** [review]
Nice!  Thanks for doing this.

**Comment by cclauss** [comment]
@Martmists-GH Can you please rename
audio_filters/equal_loudness_filter.py.broken.txt --> audio_filters/equal_loudness_filter.py
and fix the failing test?

**Comment by Martmists-GH** [comment]
I'm not sure why it's failing, it was working fine when merged. Could you provide the error it gives?

**Comment by cclauss** [comment]
I bet it is a change in numpy...
```
=================================== FAILURES ===================================
__ [doctest] audio_filters.equal_loudness_filter.EqualLoudnessFilter.process ___
053 
054         Process a single sample through both filters
055 
056         >>> filt = EqualLoudnessFilter()
UNEXPECTED EXCEPTION: AttributeError("module 'numpy' has no attribute 'int'")
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/doctest.py", line 1350, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest audio_filters.equal_loudness_filter.EqualLoudnessFilter.process[0]>", line 1, in <module>
  File "/home/runner/work/Python/Python/audio_filters/equal_loudness_filter.py", line 49, in __init__
    ya, yb = yulewalk(10, freqs_normalized, gains_normalized)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/yulewalker/yulewalker.py", line 153, in yulewalk
    j = np.arange(nb, ne).astype(np.int)
                                 ^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/numpy/__init__.py", line 284, in __getattr__
    raise AttributeError("module {!r} has no attribute "
AttributeError: module 'numpy' has no attribute 'int'
/home/runner/work/Python/Python/audio_filters/equal_loudness_filter.py:56: UnexpectedException
```

**Comment by Martmists-GH** [comment]
It looks like the `yulewalker` package is not compatible with the new numpy version for 3.11, and you should open an issue on their repository instead.

**Comment by cclauss** [comment]
I think we did... mmxgn/yulewalker#5

**Pull request merged**